### ISSUE TITLE
Consider making `cuda::complex` an aggregate

### ISSUE THREAD

**Issue by jrhemstad** [NVIDIA/libcudacxx#380]
Maybe even better, make `cuda::complex` an aggregate, with pseudo-hidden `_re`, `_im` members. (Like `std::array`)

_Originally posted by @correaa in https://github.com/NVIDIA/libcudacxx/issues/248#issuecomment-1452761890_

**Comment by jrhemstad** [comment]
We can do different things with `cuda::complex` than `cuda::std::complex`, but making it an aggregate sounds like it'd be a pretty drastic deviation. 

@correaa What would the benefit of making it an aggregate over just making it trivially default constructible and trivially copyable?

**Comment by miscco** [comment]
I want to stress that making complex an aggregate will break existing code that relies on implicit conversions from `T`. So from my point of view this is not possible

**Comment by correaa** [comment]
@miscco , There are two separate issues, backward compatibility and the problem of conversion from `T`. 

1) I understand from the thread that `cuda::complex` doesn't need to be backward compatible. i.e. minor differences in the constructor could exists.

2) I was worried about the same regarding conversions from `T`, but it is not the case according to what I understand from discussions of the standard.

I think your concern is that `cuda::complex<double> T{5.0};` may **not** initialize to the "right" complex , 5.0 + I*0.0. 
This is not the case, the trick is that aggregates will never "partially" initialize the element, if one element is initialized the others will be initialized to zero. (I didn't know that until recently.)

So, an aggregate will do the right thing, fortunately.
This in itself is not an argument to make the type an aggregate, but it to say that the aggregate is a priori compatible with conversions from `T`.

Is this your concern? If not, what is the concern?

Finally, without making any claim of code quality, here is a concrete aggregate complex that I implemented experimentally https://gitlab.com/correaa/boost-multi/-/blob/master/include/multi/adaptors/complex.hpp#L73-156 .

**Comment by jrhemstad** [comment]
> 1. I understand from the thread that `cuda::complex` doesn't need to be backward compatible. i.e. minor differences in the constructor could exists.

This is correct. `cuda::complex` is allowed to have different, breaking behavior vs `cuda::std::complex`. It's just a question of _how_ different we want it to be.

**Comment by correaa** [comment]
> What would the benefit of making it an aggregate over just making it trivially default constructible and trivially copyable?

Maybe the correct way to ask the question is why *not* make it an aggregate.

1) There are no invariants to maintain
1bis) real part and imaginary part can be accessed independently
2) The type will be memcopyable/trivial/uninitialized/partially-formed without a doubt.
3) Aggregate initialization would already do the right thing, `cuda::complex<double> z{5.1}; /* = 5.1 + i0.0*/`
4) There would be no question of whether to pass the constructor argument by reference or by value.
5) [arguably] `cuda::complex<double&>` can be instantiated, to describe complex references where the real and imaginary parts are not contiguous (this happens with some Fourier transform)
6) The class will be friendly to reflection without additional tricks
6bis) structured bindings would work out of the box `auto [re, im] = z;`.
7) real and imaginary parts will have well-defined pointers to member offsets and would be easily programmable to interact with legacy code.
8) An array library could make two real arrays (from the real and imaginary) parts without resorting to `reinterpret_cast` to skeleton classes (that end up looking like an aggregate complex itself).
9) less functions (constructors) to add `__host__ __device__ constexpr noexcept`

What is the price to pay to make it an aggregate? just that `cuda::complex<double> z(5.1, 6.7)` will not work, which IMO is not a good notation anyway. 
Being an aggregate doesn't prevent to the definition of literals or constants to allow (and encourage) better options, such as `cuda::complex<double> = 5.1 + I*6.7` or ` = 5.1 + 6.1_i`.
(When in doubt, remember that `std::array` is an aggregate.)

Note that I am not dreaming about the ideal mathematical type for complex, I am talking about a type that is fundamentally for numeric computation that can compete with numerical (e.g. Fortran) code or manually rolled complex in C.
